### PR TITLE
[Build] Fix Dependency Issues (Restore & fix after revert)

### PR DESCRIPTION
**Pull request by MichaelYochpaz**
## Related Issues
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-4543)

## Description

> Currently, in the pre_update check that installs packs from master and assures they're installed properly, the required dependencies are taken from master (API call to marketplace), while the check whether each pack is deprecated, is done locally from the branch.
> This causes issues when deprecating a pack and changing dependencies on the same PR.
> 
> This PR fixes the behavior to get deprecation data from the same API response used for fetching the dependencies instead of checking locally.

This PR is a restore & fix for https://github.com/demisto/content/pull/27405, after it has been reverted on https://github.com/demisto/content/pull/28449 (5896217e5bc2e4aeea327a288d416e647bda2af2).

### Changes in this PR
The issue and the reason that the previous changes failed on our nightly run, is that we assumed that the main method for these changes - `search_pack_and_its_dependencies`, only receives IDs of "valid" packs that aren't deprecated (hidden) when running on pre-update mode (since there we're running the packs on master without any changes).

Turns out that it does receive deprecated packs in some cases, causing the Marketplace API that's used later in the code to fetch information about its dependencies, to return an error ("Item not found") since this pack doesn't exist in there.

Since we can't check the pack-metadata file locally (since the build might be running on a branch that isn't master), and can't use the Marketplace API to receive data (it includes deprecation status), the solution that was used is to fetch the pack-metadata file from our internal GitLab using the API and a commit hash of master when the build started (we don't want to just use 'master' since it can it changes while the build runs).

Nightly run for current branch:
https://code.pan.run/xsoar/content/-/pipelines/5939489

### CODE REVIEW

**Comment by MichaelYochpaz** [comment]
Changes specific for this fix PR (without the "revert" changes, which restore the original  Marketplace changes) can be seen on:
https://github.com/demisto/content/pull/28469/files/90ae84763823c534958c859a0e6cdd6fc3011ff9..3c447bf0bbbcc5307b043be811858b569c559aa4

**Comment by MichaelYochpaz** [comment]
@daryakoval @yaakovpraisler @ilaner PR is ready for review.

Changes specific for this fix PR (without the "revert" changes, which restore the original Marketplace changes) can be seen on:
https://github.com/demisto/content/pull/28469/files/05f51e4cf33ae12ec5fce7dba008a7d829321510..48d7d77ae3b162fa52fc39f860d162582f396ddc

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 1]
```diff
@@ -286,7 +286,8 @@ def reset_core_pack_version(client: demisto_client, unremovable_packs: list):
     _, success = search_and_install_packs_and_their_dependencies(pack_ids=unremovable_packs,
                                                                  client=client,
                                                                  hostname=host,
-                                                                 install_packs_one_by_one=True)
+                                                                 multithreading=False,
+                                                                 is_post_update=False)
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_uninstall_pack.py:290]:
```suggestion
                                                                 production_bucket=True)
```

File: Tests/Marketplace/configure_and_install_packs.py [hunk 2]
```diff
@@ -37,18 +37,21 @@ def options_handler():
...
 
     for server in servers:
         logging.info(f'Starting to install all content packs in {hostname if hostname else server.internal_ip}')
-        _, success = search_and_install_packs_and_their_dependencies(pack_ids, server.client, hostname,
-                                                                     install_packs_one_by_one)
+        _, success = search_and_install_packs_and_their_dependencies(pack_ids=pack_ids,
+                                                                     client=server.client,
+                                                                     hostname=hostname,
+                                                                     multithreading=use_multithreading,
+                                                                     is_post_update=True)
```
Comment (daryakoval) [inline Tests/Marketplace/configure_and_install_packs.py:54]:
```suggestion
                                                                    production_bucket=False)
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 3]
```diff
@@ -692,51 +729,67 @@ def search_and_install_packs_and_their_dependencies_private(test_pack_path: str,
 
 def search_and_install_packs_and_their_dependencies(pack_ids: list,
                                                     client: demisto_client, hostname: str | None = None,
-                                                    install_packs_one_by_one=False):
-    """ Searches for the packs from the specified list, searches their dependencies, and then
+                                                    multithreading: bool = True,
+                                                    is_post_update: bool = False):
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:733]:
```suggestion
                                                    production_bucket: bool = True):
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 4]
```diff
@@ -692,51 +729,67 @@ def search_and_install_packs_and_their_dependencies_private(test_pack_path: str,
...
     Args:
         pack_ids (list): A list of the pack ids to search and install.
         client (demisto_client): The client to connect to.
         hostname (str): Hostname of instance. Using for logs.
-        install_packs_one_by_one(bool): Whether to install packs in small batches.
-            If false - install all packs in one batch.
-
+        multithreading (bool): Whether to use multithreading to install packs in parallel.
+            If multithreading is used, installation requests will be sent in batches of each pack and its dependencies.
+        is_post_update (bool): Whether the installation is in post update mode. Defaults to False.
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:744]:
```suggestion
        production_bucket (bool): Whether the installation is from the production bucket or the custom bucket. Defaults to True.
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 5]
```diff
@@ -692,51 +729,67 @@ def search_and_install_packs_and_their_dependencies_private(test_pack_path: str,
...
+        'commit_hash': master_commit_hash,
+    }
+
+    if is_post_update:
+        logging.info("Detected post-update run mode. "
+                     "Pack deprecation status will be determined using local pack metadata.")
+
+    else:
+        logging.info("Detected pre-update run mode. "
+                     "Pack deprecation status will be determined using Marketplace API.")
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:783]: Remove from here, if you want to log this, add this to the `is_pack_deprecated` function, and use `logging.debug` not info. 
```suggestion
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 6]
```diff
@@ -692,51 +729,67 @@ def search_and_install_packs_and_their_dependencies_private(test_pack_path: str,
...
     lock = Lock()
 
-    if install_packs_one_by_one:
+    kwargs = {
+        'client': client,
+        'packs_to_install': packs_to_install,
+        'installation_request_body': installation_request_body,
+        'lock': lock,
+        'collected_dependencies': collected_dependencies,
+        'is_post_update': is_post_update,
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:771]:
```suggestion
        'production_bucket': production_bucket,
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 7]
```diff
@@ -453,76 +441,125 @@ def search_pack_and_its_dependencies(client: demisto_client,
                                      packs_to_install: list,
                                      installation_request_body: list,
                                      lock: Lock,
-                                     packs_in_the_list_to_install: list,
-                                     one_pack_and_its_dependencies_in_batch: bool = False,
-                                     batch_packs_install_request_body: list = None,
+                                     collected_dependencies: list,
+                                     is_post_update: bool,
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:445]:
```suggestion
                                     production_bucket: bool,
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 8]
```diff
@@ -453,76 +441,125 @@ def search_pack_and_its_dependencies(client: demisto_client,
...
     Args:
         client (demisto_client): The configured client to use.
         pack_id (str): The id of the pack to be installed.
         packs_to_install (list) A list of the packs to be installed in this iteration.
         installation_request_body (list): A list of packs to be installed, in the request format.
         lock (Lock): A lock object.
-        packs_in_the_list_to_install (list): list of packs that are already in the list to install
-        one_pack_and_its_dependencies_in_batch(bool): Whether to install packs in small batches.
+        collected_dependencies (list): list of packs that are already in the list to install
+        is_post_update (bool): Whether the installation is done in post-update or not (pre-update otherwise).
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:462]:
```suggestion
        production_bucket (bool): Whether the installation is done using production bucket or custom bucket.
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 9]
```diff
@@ -453,76 +441,125 @@ def search_pack_and_its_dependencies(client: demisto_client,
...
+
+    if dependencies_data:
+        dependencies_ids = [dependency['id'] for dependency in dependencies_data]
+        logging.debug(f"Found dependencies for '{pack_id}': {dependencies_ids}")
+
+        for dependency in dependencies_data:
+            dependency_id = dependency['id']
+            # If running on pre-update, we check for deprecation using API data.
+            # if running on post-update, we check for deprecation locally on the branch.
+            is_deprecated = is_pack_deprecated(pack_id=dependency_id, check_locally=is_post_update, pack_api_data=dependency)
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:517]:
```suggestion
            is_deprecated = is_pack_deprecated(pack_id=dependency_id, check_locally=not production_bucket, pack_api_data=dependency)
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 10]
```diff
@@ -453,76 +441,125 @@ def search_pack_and_its_dependencies(client: demisto_client,
...
+                                       checked_packs=[pack_id])
+
+    if dependencies_data:
+        dependencies_ids = [dependency['id'] for dependency in dependencies_data]
+        logging.debug(f"Found dependencies for '{pack_id}': {dependencies_ids}")
+
+        for dependency in dependencies_data:
+            dependency_id = dependency['id']
+            # If running on pre-update, we check for deprecation using API data.
+            # if running on post-update, we check for deprecation locally on the branch.
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:516]: remove from here
```suggestion
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 11]
```diff
@@ -33,233 +35,215 @@
...
+
+    file_data_b64 = response.json()['content']
+    file_data = base64.b64decode(file_data_b64).decode('utf-8')
 
+    return json.loads(file_data)
 
-def get_pack_display_name(pack_id: str) -> str:
+
+def is_pack_deprecated(pack_id: str, check_locally: bool = True,
+                       commit_hash: str | None = None, pack_api_data: dict | None = None) -> bool:
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:91]: Please change the function, move all logic related to the deprecated check inside this function, using my other comments, and add the explanations for the docstrings here.

**Review by daryakoval — CHANGES_REQUESTED** [review]
Great, lets change the logic a little, please see my comments.
Feel free to contact me if any assistance is needed.

File: Tests/Marketplace/search_and_install_packs.py [hunk 12]
```diff
@@ -33,233 +35,215 @@
...
-                dependencies_data.append({
-                    'id': dependency.get('id'),
-                    'version': dependency.get('extras', {}).get('pack', {}).get('currentVersion')
-                })
-                next_call_dependants_ids.append(dependency.get('id'))
-                checked_packs.append(dependency.get('id'))
+            if dependant in dependants_ids and is_required and dependency['id'] not in checked_packs:
+                dependencies_data.append(dependency)
+                next_call_dependants_ids.append(dependency['id'])
+                checked_packs.append(dependency['id'])
```
[the remark this answers is not in the document]
Comment (MichaelYochpaz) [reply Tests/Marketplace/search_and_install_packs.py:176]: This is data that's returned from the Marketplace API. There shouldn't be a server response with pack data, but not an ID. If the API does change, functions called before this one will break anyway.

File: Tests/Marketplace/search_and_install_packs.py [hunk 13]
```diff
@@ -453,76 +458,119 @@ def search_pack_and_its_dependencies(client: demisto_client,
...
-                    {p['id']: p for p in current_packs_to_install if p['id'] not in packs_in_the_list_to_install}
-                if pack_and_its_dependencies:
-                    packs_in_the_list_to_install += pack_and_its_dependencies
-                    batch_packs_install_request_body.append(list(pack_and_its_dependencies.values()))  # type:ignore[union-attr]
+    global SUCCESS_FLAG
+    is_deprecated = is_pack_deprecated(pack_id=pack_id, production_bucket=production_bucket, commit_hash=commit_hash)
+
+    if is_deprecated:
+        logging.warning(f"Pack '{pack_id}' is deprecated (hidden) and will not be installed.")
+        return
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:496]:
```suggestion
    if is_pack_deprecated(pack_id=pack_id, production_bucket=production_bucket, commit_hash=commit_hash):
        logging.warning(f"Pack '{pack_id}' is deprecated (hidden) and will not be installed.")
        return
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 14]
```diff
@@ -453,76 +458,119 @@ def search_pack_and_its_dependencies(client: demisto_client,
...
+
+    api_data = get_pack_dependencies(client=client, pack_id=pack_id, lock=lock)
+
+    if not api_data:
+        return  # Error information already logged on 'get_pack_dependencies'.
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:501]:
```suggestion
    if not api_data:
        return
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 15]
```diff
@@ -453,76 +458,119 @@ def search_pack_and_its_dependencies(client: demisto_client,
...
+        pack_api_data = api_data['packs'][0]
+        current_packs_to_install = [pack_api_data]
+
+        create_dependencies_data_structure(response_data=api_data.get('dependencies', []),
+                                           dependants_ids=[pack_id],
+                                           dependencies_data=dependencies_data,
+                                           checked_packs=[pack_id])
+
+    except Exception as ex:
+        # If APIs response structure has unexpectedly changed, causing an error, print debug information
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:515]:
```suggestion
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 16]
```diff
@@ -692,51 +740,62 @@ def search_and_install_packs_and_their_dependencies_private(test_pack_path: str,
...
+        'multithreading': multithreading,
+        'batch_packs_install_request_body': batch_packs_install_request_body,
+        'commit_hash': master_commit_hash,
+    }
+
+    if production_bucket:
+        logging.debug("Packs deprecation status will be determined by their status on our production bucket.")
+
+    else:
+        logging.debug("Packs deprecation status will be determined by the local branch and its changes.")
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:789]: Please delete. Explain these differences in the `is_pack_deprecated` docstring.
```suggestion
```

**Review by daryakoval — CHANGES_REQUESTED** [review]
Great work! Please see my notes. I don't support adding to much comments, since this codes goes trough refactoring a lot, the code changes but the comments stays and become wrong. 

Please trigger nightly and test upload flow, update me when they pass and I will approve the PR.

**Comment by MichaelYochpaz** [comment]
Updated nightly: https://code.pan.run/xsoar/content/-/pipelines/5946644

**Review by daryakoval — APPROVED** [review]
Let the build be on our side!

**Pull request merged**